### PR TITLE
CCP-67: Make userId optional for engage sms and email

### PR DESCRIPTION
**Pull request by zawadzkip**
Making the userId optional because the new test message flow in engage will no longer have access to this. We are removing functionality where traits are populated in a message from a user profile as this can lead to a PII leak.

The original functionality will still exist, however our frontend will no longer be able to retrieve a `profileId` which we use as input for a `userId` into this action. I have added an error for better visibility on a missing `userId` with `traitEnrichment` disabled.

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/engage-messaging-twilio/sendSms/sms-sender.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ export class SmsMessageSender extends MessageSender<Payload> {
     }
 
     const profile = {
-      user_id: this.payload.userId,
+      user_id: this.payload.userId || undefined,
```
Comment (ryanrouleau) [inline packages/destination-actions/src/destinations/engage-messaging-twilio/sendSms/sms-sender.ts:33]: same: `??`

File: packages/destination-actions/src/destinations/engage-messaging-twilio/sendSms/sms-sender.ts [hunk 2]
```diff
@@ -54,6 +54,13 @@ export class SmsMessageSender extends MessageSender<Payload> {
 
   private getProfileTraits = async () => {
     try {
+      if (!this.payload.userId) {
```
Comment (alfrimpong) [inline packages/destination-actions/src/destinations/engage-messaging-twilio/sendSms/sms-sender.ts:57]: does this have to be inside the try block?

Comment (zawadzkip) [reply packages/destination-actions/src/destinations/engage-messaging-twilio/sendSms/sms-sender.ts:57]: no not necessarily, ends up removing the check in the catch block then, I will update that. thanks.

File: packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts [hunk 3]
```diff
@@ -430,7 +437,7 @@ const action: ActionDefinition<Settings, Payload> = {
                 ...payload.customArgs,
                 source_id: settings.sourceId,
                 space_id: settings.spaceId,
-                user_id: payload.userId,
+                user_id: payload.userId ?? undefined,
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts:440]: Is coalescing operator required here given that the type of payload.userId is string or undefined (userId?) ?

File: packages/destination-actions/src/destinations/engage-messaging-twilio/sendSms/sms-sender.ts [hunk 4]
```diff
@@ -30,7 +30,7 @@ export class SmsMessageSender extends MessageSender<Payload> {
     }
 
     const profile = {
-      user_id: this.payload.userId,
+      user_id: this.payload.userId ?? undefined,
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/engage-messaging-twilio/sendSms/sms-sender.ts:33]: Same as above. Is coalescing operator required here given that the type of payload.userId is string or undefined (userId?) ?

**Review by varadarajan-tw — APPROVED** [review]
LGTM! Left a comment but thats not a blocker.

File: packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts (hunk 3 above)
Comment (zawadzkip) [reply packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts:440]: its really just as a backup, I suppose someone could pass in an empty string here but that's the only other situation I can imagine would be relevant for the coalescing operator.

**Pull request merged**